### PR TITLE
SCRD-4496 build from HEAD to avoid older packages for ardana

### PR DESCRIPTION
**Pull request by JanZerebecki**
Packages in the mirrored cloud installation image might be a bit older
due to mirror lag. The repo from the patch under test and its declared
dependencies are build from Git. Also building the other Ardana packages
from Git ensures that the mirror lag doesn't affect them.

Skip packages whose git commit is the same as what is requested to avoid
rebuilding them.

### CODE REVIEW

File: scripts/jenkins/ardana/gerrit/build_test_package.py [hunk 1]
```diff
@@ -470,6 +473,13 @@ def build_test_packages(change_ids, obs_linked_project, home_project,
             change_ids.extend(
                 packages[c.gerrit_project].get_depends_on_changes())
 
+    for project_name, package in gerrit_project_map().items():
```
Comment (stefannica) [inline scripts/jenkins/ardana/gerrit/build_test_package.py]: This approach will duplicate all source tracking Ardana packages in the test IBS project. One way to avoid the duplication and the additional time spent building all packages is to use the Devel:Cloud:X:Staging:Ardana IBS project, which has package publishing enabled. We could configure this repository in addition to the one corresponding to the test IBS project, but we'd also need to protect against the case in which the packages in the D:C:X:S:Ardana repository are more recent than the ones in the test repository and override them. This corner case would be really bad, because the test will run against the D:C:X:S:Ardana packages instead of the Gerrit patched packages and will report false positives.

Comment (JanZerebecki) [reply scripts/jenkins/ardana/gerrit/build_test_package.py]: I added skipping if its already the same. Will this work sufficiently?

Comment (stefannica) [reply scripts/jenkins/ardana/gerrit/build_test_package.py]: > I added skipping if its already the same. Will this work sufficiently?

This won't make any difference regarding my objections. Most Gerrit changes only target one repository and one package, so most of the time you will end up rebuilding 41 packages that are already available in `Devel:Cloud:9:Staging:Ardana`. 

The correct solution to this problem is to configure `http://download.suse.de/ibs/Devel:/Cloud:/9:/Staging:/Ardana/SLE_12_SP4/` as a repository in the Ardana deployer, but also make sure that the packages in the test repository will always be selected first (so it should have a higher priority). We can already supply additional repositories using the `extra_repos` openstack-ardana parameter [2], but that doesn't cover setting the priority. Maybe you could extend that support to allow specifying the priority as an optional suffix (e.g. <repository-url>[:<repository-prio>]).

On the other hand, the gap that you're trying to close isn't that great a problem anymore, because according to the infrastructure doc [1], the Ardana packages are sync-ed to Provo every hour or so, and the explicit Depends-On changes are already picked up by the test package builder, so the only scenario that could still affect us is the one with implicit dependencies, which 1. isn't really that frequent, and 2. might do us more good than harm, because it will reveal these implicit dependencies and maybe force the developer to specify them as explicit Depends-Ons (this information is really useful during MU release, which @rsalevsky will confirm). Even so, I would like to see it closed.

[1] https://github.com/SUSE/cloud/blob/master/docs/CI/ci-infrastructure.rst#package-mirroring
[2] https://github.com/SUSE-Cloud/automation/blob/master/jenkins/ci.suse.de/templates/cloud-ardana-pipeline-template.yaml#L247

Comment (JanZerebecki) [reply scripts/jenkins/ardana/gerrit/build_test_package.py]: > The correct solution to this problem is to configure http://download.suse.de/ibs/Devel:/Cloud:/9:/Staging:/Ardana/SLE_12_SP4/ as a repository in the Ardana deployer

This is not necessary as the ardana-ci project used by this script already inherits from Devel:Cloud:9:Staging.

OTOH we discussed in chat https://chat.suse.de/channel/cloud-ci-squad?msg=HZxQHLLukjaZYSq5k that we will move everything to base on Devel:Cloud:9 without Staging and just rebuild a few more packages that are different on master. But that won't change anything in this PR.

**Comment by JanZerebecki** [comment]
Rebased.
